README.md
<div align="center">
<h1>Sticky Item</h1>

[![npm](https://badgen.net/npm/v/@gorhom/sticky-item)](https://www.npmjs.com/package/@gorhom/sticky-item) [![npm](https://badgen.net/npm/license/@gorhom/sticky-item)](https://www.npmjs.com/package/@gorhom/sticky-item) [![npm](https://badgen.net/npm/types/@gorhom/sticky-item)](https://www.npmjs.com/package/@gorhom/sticky-item)

<img src="./preview.gif">

An interactive sticky item inspired by Facebook Stories.

</div>

---

## Installation

```sh
yarn add @gorhom/sticky-item
# or
npm install @gorhom/sticky-item
```

> Also, you need to install [react-native-reanimated](https://github.com/software-mansion/react-native-reanimated), [react-native-gesture-handler](https://github.com/software-mansion/react-native-gesture-handler) & [react-native-svg](https://github.com/react-native-community/react-native-svg), and follow theirs installation instructions.

## Usage

```tsx
...
import StickyItemFlatList from '@gorhom/sticky-item';

// dummy data
const data = [...Array(20)]
  .fill(0)
  .map((_, index) => ({ id: `item-${index}` }));

// configs
const ITEM_WIDTH = 90;
const ITEM_HEIGHT = 150;
const STICKY_ITEM_WIDTH = 24;
const STICKY_ITEM_HEIGHT = 24;
const STICKY_ITEM_BACKGROUNDS = ['#222', '#000'];
const SEPARATOR_SIZE = 8;
const BORDER_RADIUS = 10;

const StickyItemView = ({
  x,
  threshold,
  itemWidth,
  itemHeight,
  stickyItemWidth,
  stickyItemHeight,
  separatorSize,
  isRTL,
}) => {

  const amazingAnimation = {
    // here you add your custom interactive animation
    // based on the animated value `x`
  }

  return <Animated.View style={amazingAnimation} />
}

const App = () => {
  // methods
  const handleStickyItemPress = () => Alert.alert('Sticky Item Pressed');

  // render
  const renderItem = ({ item, index}) => (
    <View
      key={`item-${index}`}
      style={{
        backgroundColor: 'red',
        width: ITEM_WIDTH,
        height: ITEM_HEIGHT,
      }}
    />
  )
  return (
    <StickyItemFlatList
      itemWidth={ITEM_WIDTH}
      itemHeight={ITEM_HEIGHT}
      separatorSize={SEPARATOR_SIZE}
      borderRadius={BORDER_RADIUS}
      stickyItemWidth={STICKY_ITEM_WIDTH}
      stickyItemHeight={STICKY_ITEM_HEIGHT}
      stickyItemBackgroundColors={STICKY_ITEM_BACKGROUNDS}
      stickyItemContent={StickyItemView}
      onStickyItemPress={handleStickyItemPress}
      data={data}
      renderItem={renderItem}
    />
  )
}

export default App
```

## Props

| name                         | description                                                                                                   | required | type                                                | default |
| ---------------------------- | ------------------------------------------------------------------------------------------------------------- | -------- | --------------------------------------------------- | ------- |
| `itemWidth`                  | Item's width.                                                                                                 | YES      | number                                              |         |
| `itemHeight`                 | Item's height.                                                                                                | YES      | number                                              |         |
| `separatorSize`              | FlatList's separator width, \* if you provide `ItemSeparatorComponent`, you will need to set `separatorSize`. | NO\*     | number                                              | 10      |
| `borderRadius`               | Item & sticky border radius.                                                                                  | NO       | number                                              | 15      |
| `stickyItemActiveOpacity`    | Sticky item's active opacity.                                                                                 | YES      | number                                              | 0.25    |
| `stickyItemWidth`            | Sticky item's width.                                                                                          | YES      | number                                              |         |
| `stickyItemHeight`           | Sticky item's height.                                                                                         | YES      | number                                              |         |
| `stickyItemBackgroundColors` | Sticky item's two background colors, one when sticky item is extended another when it's minimize.             | YES      | string[]                                            |         |
| `stickyItemContent`          | Sticky item's content component.                                                                              | YES      | [`ReactNode`](./src/types.ts#L30)                   |         |
| `isRTL`                      | FlatList' layout direction.                                                                                   | NO       | boolean                                             | false   |
| `onStickyItemPress`          | Callback when sticky item gets pressed.                                                                       | NO       | function                                            |         |
| `...FlatList Props`          | React Native FlatList props.                                                                                  | NO       | [`FlatList`](https://reactnative.dev/docs/flatlist) |         |

## To Do

- [ ] Write a detailed step-by-step instruction to create sticky interactive animation.
- [ ] Add more examples.
- [ ] Add vertical support ?.

<h2 id="built-with">Built With ❤️</h2>

- [react-native-reanimated](https://github.com/software-mansion/react-native-reanimated)
- [react-native-gesture-handler](https://github.com/software-mansion/react-native-gesture-handler)
- [react-native-redash](https://github.com/wcandillon/react-native-redash)
- [react-native-svg](https://github.com/react-native-community/react-native-svg)
- [@react-native-community/bob](https://github.com/react-native-community/bob)

## Author

- [Mo Gorhom](https://twitter.com/gorhom)

## License

MIT

<div align="center">

Liked the library? 😇

<a href="https://www.buymeacoffee.com/gorhom" target="_blank"><img src="https://cdn.buymeacoffee.com/buttons/default-red.png" alt="Buy Me A Coffee" height="50" ></a>

</div>

---

<p align="center">
<a href="https://twitter.com/gorhom"><img src="./logo.png"></a>
</p>


CHANGELOG.md
# Changelog

## [v2.1.1](https://github.com/gorhom/react-native-sticky-item/compare/v2.1.0...v2.1.1)

#### Improvements

- refactor: copy redash used functionalities ([`#24`](https://github.com/gorhom/react-native-sticky-item/pull/24)).
- chore: updated dependencies ([`#23`](https://github.com/gorhom/react-native-sticky-item/pull/23)).

## [v2.1.0](https://github.com/gorhom/react-native-sticky-item/compare/v2.0.0...v2.1.0) - 2021-01-19

#### Features

- feat: added reanimated 2 support ([`#19`](https://github.com/gorhom/react-native-sticky-item/pull/19)).

#### Improvements

- chore: updated dependencies ([1fbca83](https://github.com/gorhom/react-native-sticky-item/commit/1fbca835538d4f280da161ffa0f16c69eb56a3de)).
- chore: updated examples styling ([4b744ea](https://github.com/gorhom/react-native-sticky-item/commit/4b744ea96f6a1b5613203a7c208d714acb7762ca)).
- chore: added auto-changelog config ([ed464a0](https://github.com/gorhom/react-native-sticky-item/commit/ed464a0d61b880cd3dd83a3e8d11b34ad5fb87c6)).
- refactor: updated animated thread to be UI instead of native ([b6d7499](https://github.com/gorhom/react-native-sticky-item/commit/b6d74993eca2d0f02b7029aecba44edc39de296a)).

## [v2.0.0](https://github.com/gorhom/react-native-sticky-item/compare/v1.3.2...v2.0.0) - 2020-06-04

#### Improvements

- refactor: updated pointerevents handling (#15) ([ff49f5e](https://github.com/gorhom/react-native-sticky-item/commit/ff49f5ea26f44a60bbb1f3531e3b8f75f7281bad)).

#### Fixes

- fix: sticky item width calculation ([`#14`](https://github.com/gorhom/react-native-sticky-item/pull/14)).
- fix: fixed pointerevents on android (#15) ([1295c95](https://github.com/gorhom/react-native-sticky-item/commit/1295c953bfa4cdce0e8eed882ffe7f556990930e)).

## [v1.3.2](https://github.com/gorhom/react-native-sticky-item/compare/v1.3.1...v1.3.2) - 2020-06-03

#### Fixes

- fix: updated sticky item pointerevents to solve pressable area (#15) ([1e0af37](https://github.com/gorhom/react-native-sticky-item/commit/1e0af378ea41aaafa36bc87a3d716067cf7f0d79)).

## [v1.3.1](https://github.com/gorhom/react-native-sticky-item/compare/v1.3.0...v1.3.1) - 2020-06-01

#### Fixes

- Revert "fix: sticky item width calculation (#13)"

This reverts commit e5b246b54604d3c2bed0dc675cf6df483182e8cc. ([206420f](https://github.com/gorhom/react-native-sticky-item/commit/206420f295a9d1f84914ba13586e146ecb2dd03e)).

## [v1.3.0](https://github.com/gorhom/react-native-sticky-item/compare/v1.2.1...v1.3.0) - 2020-06-01

#### Features

- feat: allow custom separator, optimise current one by @Hirbod (#9) ([`#11`](https://github.com/gorhom/react-native-sticky-item/pull/11)).

#### Improvements

- chore: add sticky item active effect ([`#10`](https://github.com/gorhom/react-native-sticky-item/pull/10)).
- chore: allow decelerationRate to be overrided ([c0e2359](https://github.com/gorhom/react-native-sticky-item/commit/c0e23590dbc440e44a468ebb5fde2a5db9f616da)).

#### Fixes

- fix: sticky item width calculation ([`#13`](https://github.com/gorhom/react-native-sticky-item/pull/13)).

## [v1.2.1](https://github.com/gorhom/react-native-sticky-item/compare/v1.2.0...v1.2.1) - 2020-05-28

#### Fixes

- fix: fixed the hitslop and pointerevent for android ([7062ab9](https://github.com/gorhom/react-native-sticky-item/commit/7062ab9cc97d8352ef26e2e5a1dee04081110bbe)).

## [v1.2.0](https://github.com/gorhom/react-native-sticky-item/compare/v1.1.2...v1.2.0) - 2020-05-28

#### Features

- feat: handle initial index ([`#6`](https://github.com/gorhom/react-native-sticky-item/pull/6)).

## [v1.1.2](https://github.com/gorhom/react-native-sticky-item/compare/v1.1.1...v1.1.2) - 2020-05-25

#### Improvements

- chore: forward FlatList ref ([`#3`](https://github.com/gorhom/react-native-sticky-item/pull/3)).

## [v1.1.1](https://github.com/gorhom/react-native-sticky-item/compare/v1.1.0...v1.1.1) - 2020-05-03

#### Improvements

- chore: downgrade autochangelog ([0378b97](https://github.com/gorhom/react-native-sticky-item/commit/0378b97a3a58819dd25a15cb0e9be69af7df560f)).

## v1.1.0 - 2020-05-03

#### Features

- feat: added RTL support ([`#1`](https://github.com/gorhom/react-native-sticky-item/pull/1)).

#### Improvements

- chore: updated example icon ([50f92b5](https://github.com/gorhom/react-native-sticky-item/commit/50f92b59c6c87b93138980c6ef5561a18e7a74df)).
- chore: added basic example ([e3e7169](https://github.com/gorhom/react-native-sticky-item/commit/e3e71691dd91a51d7997b4d8a599e54671eee693)).
- chore: added onStickyItemPress callback ([8c248b5](https://github.com/gorhom/react-native-sticky-item/commit/8c248b5aba7d125d1e46e7bbc0bff78c741ab1d1)).
- refactor: updated examples ([4fbb383](https://github.com/gorhom/react-native-sticky-item/commit/4fbb38386fbee2ef36289a7a0ca958a48172345a)).
- chore: updated default props ([ff98c99](https://github.com/gorhom/react-native-sticky-item/commit/ff98c997f753ba8093db62873c8c42c54adab936)).
- refactor: updated basic rtl example ([0c499ec](https://github.com/gorhom/react-native-sticky-item/commit/0c499ec0e8024c91ac0b8f64b9ef2985e4c4d73b)).
- chore: updated examples styling ([08decb0](https://github.com/gorhom/react-native-sticky-item/commit/08decb02208b69ba10bdc1e77e8bcfbcac186276)).
- chore: added commitlint/cli ([d8df485](https://github.com/gorhom/react-native-sticky-item/commit/d8df485b9dfbb5c4783ff21c9426269fa08d26cf)).
- chore: added auto-changelog ([ddd60cf](https://github.com/gorhom/react-native-sticky-item/commit/ddd60cff1536272c03cc649a17e1f63065366bcb)).
- chore: fix tsc warning ([4b3fe29](https://github.com/gorhom/react-native-sticky-item/commit/4b3fe2911323a2f516acf3d3795ed6423b875a0a)).

#### Fixes

- fix: fixed tapgesture hitslop ([139f2d3](https://github.com/gorhom/react-native-sticky-item/commit/139f2d312fb841c54f97cd1c106f3bf1bc1c4aac)).

#### Documentations


src/utilities/redash.ts
/**
 * MIT License
 * Copyright (c) 2020 William Candillon
 * https://github.com/wcandillon/react-native-redash
 */

import { useRef } from 'react';
import { processColor } from 'react-native';
import Animated from 'react-native-reanimated';
// @ts-ignore
import parseSVG from 'parse-svg-path';
// @ts-ignore
import absSVG from 'abs-svg-path';
// @ts-ignore
import normalizeSVG from 'normalize-svg-path';
import { cubicBezierLength } from './cubicBezierLength';
import {
  FlingGestureHandlerGestureEvent,
  ForceTouchGestureHandlerGestureEvent,
  GestureHandlerStateChangeNativeEvent,
  LongPressGestureHandlerGestureEvent,
  PanGestureHandlerGestureEvent,
  PinchGestureHandlerGestureEvent,
  RotationGestureHandlerGestureEvent,
  TapGestureHandlerGestureEvent,
} from 'react-native-gesture-handler';
import { Easing, interpolate } from './reanimated';

const {
  multiply,
  event,
  floor,
  abs,
  sub,
  block,
  Extrapolate,
  set,
  color,
  add,
  not,
  cond,
  neq,
  concat,
  startClock,
  min: min2,
  max: max2,
  stopClock,
  timing,
  round,
  proc,
  Clock,
  Value,
} = Animated;

type Color = Animated.Adaptable<string> | Animated.Adaptable<number>;

interface ColorInterpolationConfig {
  inputRange: readonly Animated.Adaptable<number>[];
  outputRange: Color[];
}

type Adaptable<T> = { [P in keyof T]: Animated.Adaptable<T[P]> };

export interface Vector<
  T extends Animated.Adaptable<number> = Animated.Adaptable<number>
> {
  x: T;
  y: T;
}

type NativeEvent = GestureHandlerStateChangeNativeEvent &
  (
    | PanGestureHandlerGestureEvent
    | TapGestureHandlerGestureEvent
    | LongPressGestureHandlerGestureEvent
    | RotationGestureHandlerGestureEvent
    | FlingGestureHandlerGestureEvent
    | PinchGestureHandlerGestureEvent
    | ForceTouchGestureHandlerGestureEvent
  );

export type TimingConfig = Partial<Omit<Animated.TimingConfig, 'toValue'>>;

type Transform2dName =
  | 'translateX'
  | 'translateY'
  | 'scale'
  | 'skewX'
  | 'skewY'
  | 'scaleX'
  | 'scaleY'
  | 'rotateZ'
  | 'rotate';

type Transformations = {
  [Name in Transform2dName]: Animated.Adaptable<number>;
};

export type Transforms2d = (
  | Pick<Transformations, 'translateX'>
  | Pick<Transformations, 'translateY'>
  | Pick<Transformations, 'scale'>
  | Pick<Transformations, 'scaleX'>
  | Pick<Transformations, 'scaleY'>
  | Pick<Transformations, 'skewX'>
  | Pick<Transformations, 'skewY'>
  | Pick<Transformations, 'rotateZ'>
  | Pick<Transformations, 'rotate'>
)[];

export const transformOrigin = (
  { x, y }: Vector,
  ...transformations: Transforms2d
): Transforms2d => [
  { translateX: x },
  { translateY: y },
  ...transformations,
  { translateX: multiply(x, -1) },
  { translateY: multiply(y, -1) },
];

type Atomic = string | number | boolean;

export const useValue = <V extends Atomic>(value: V) =>
  useConst(() => new Value(value));

export const useConst = <T>(initialValue: T | (() => T)): T => {
  const ref = useRef<{ value: T }>();
  if (ref.current === undefined) {
    // Box the value in an object so we can tell if it's initialized even if the initializer
    // returns/is undefined
    ref.current = {
      value:
        typeof initialValue === 'function'
          ? (initialValue as Function)()
          : initialValue,
    };
  }
  return ref.current.value;
};

export const onGestureEvent = (
  nativeEvent: Partial<Adaptable<NativeEvent>>
) => {
  const gestureEvent = event([{ nativeEvent }]);
  return {
    onHandlerStateChange: gestureEvent,
    onGestureEvent: gestureEvent,
  };
};

export const clamp = proc(
  (
    value: Animated.Adaptable<number>,
    lowerBound: Animated.Adaptable<number>,
    upperBound: Animated.Adaptable<number>
  ): Animated.Node<number> => min2(max2(lowerBound, value), upperBound)
);

export const useGestureHandler = (
  nativeEvent: Parameters<typeof onGestureEvent>[0]
) => useConst(() => onGestureEvent(nativeEvent));

export const withTransition = (
  value: Animated.Node<number>,
  timingConfig: TimingConfig = {}
) => {
  const init = new Value(0);
  const clock = new Clock();
  const state = {
    finished: new Value(0),
    frameTime: new Value(0),
    position: new Value(0),
    time: new Value(0),
  };
  const config = {
    toValue: new Value(0),
    duration: 150,
    easing: Easing.linear,
    ...timingConfig,
  };
  return block([
    cond(not(init), [set(init, 1), set(state.position, value)]),
    cond(neq(config.toValue, value), [
      set(state.frameTime, 0),
      set(state.time, 0),
      set(state.finished, 0),
      set(config.toValue, value),
      startClock(clock),
    ]),
    timing(clock, state, config),
    cond(state.finished, stopClock(clock)),
    state.position,
  ]);
};

export const withTimingTransition = withTransition;
export const fract = (x: Animated.Adaptable<number>) => sub(x, floor(x));

export const opacity = (c: number) => ((c >> 24) & 255) / 255;
export const red = (c: number) => (c >> 16) & 255;
export const green = (c: number) => (c >> 8) & 255;
export const blue = (c: number) => c & 255;

export const interpolateColor = (
  value: Animated.Adaptable<number>,
  config: ColorInterpolationConfig,
  colorSpace: 'hsv' | 'rgb' = 'rgb'
): Animated.Node<number> => {
  const { inputRange } = config;
  const outputRange = config.outputRange.map(c =>
    //@ts-ignore
    typeof c === 'number' ? c : processColor(c)
  );
  if (colorSpace === 'hsv') {
    //@ts-ignore
    return interpolateColorsHSV(value, inputRange, outputRange);
  }
  //@ts-ignore
  return interpolateColorsRGB(value, inputRange, outputRange);
};

export const mix = proc(
  (
    value: Animated.Adaptable<number>,
    x: Animated.Adaptable<number>,
    y: Animated.Adaptable<number>
  ) => add(x, multiply(value, sub(y, x)))
);

const rgbToHsv = (c: number) => {
  const r = red(c) / 255;
  const g = green(c) / 255;
  const b = blue(c) / 255;

  const ma = Math.max(r, g, b);
  const mi = Math.min(r, g, b);
  let h = 0;
  const v = ma;

  const d = ma - mi;
  const s = ma === 0 ? 0 : d / ma;
  if (ma === mi) {
    h = 0; // achromatic
  } else {
    switch (ma) {
      case r:
        h = (g - b) / d + (g < b ? 6 : 0);
        break;
      case g:
        h = (b - r) / d + 2;
        break;
      case b:
        h = (r - g) / d + 4;
        break;
      default: // do nothing
    }
    h /= 6;
  }
  return { h, s, v };
};

export const hsv2rgb = (
  h: Animated.Adaptable<number>,
  s: Animated.Adaptable<number>,
  v: Animated.Adaptable<number>
) => {
  // vec4 K = vec4(1.0, 2.0 / 3.0, 1.0 / 3.0, 3.0);
  const K = {
    x: 1,
    y: 2 / 3,
    z: 1 / 3,
    w: 3,
  };
  // vec3 p = abs(fract(c.xxx + K.xyz) * 6.0 - K.www);
  const p = {
    x: abs(sub(multiply(fract(add(h, K.x)), 6), K.w)),
    y: abs(sub(multiply(fract(add(h, K.y)), 6), K.w)),
    z: abs(sub(multiply(fract(add(h, K.z)), 6), K.w)),
  };
  // return c.z * mix(K.xxx, clamp(p - K.xxx, 0.0, 1.0), c.y);
  const rgb = {
    x: multiply(v, mix(s, K.x, clamp(sub(p.x, K.x), 0, 1))),
    y: multiply(v, mix(s, K.x, clamp(sub(p.y, K.x), 0, 1))),
    z: multiply(v, mix(s, K.x, clamp(sub(p.z, K.x), 0, 1))),
  };
  return {
    r: round(multiply(rgb.x, 255)),
    g: round(multiply(rgb.y, 255)),
    b: round(multiply(rgb.z, 255)),
  };
};

export const hsv2color = proc(
  // @ts-ignore
  (
    h: Animated.Adaptable<number>,
    s: Animated.Adaptable<number>,
    v: Animated.Adaptable<number>
  ) => {
    const { r, g, b } = hsv2rgb(h, s, v);
    return color(r, g, b);
  }
);

const interpolateColorsHSV = (
  animationValue: Animated.Adaptable<number>,
  inputRange: readonly Animated.Adaptable<number>[],
  colors: number[]
): Animated.Node<number> => {
  const colorsAsHSV = colors.map(c => rgbToHsv(c));
  const h = interpolate(animationValue, {
    inputRange,
    outputRange: colorsAsHSV.map(c => c.h),
    extrapolate: Extrapolate.CLAMP,
  });
  const s = interpolate(animationValue, {
    inputRange,
    outputRange: colorsAsHSV.map(c => c.s),
    extrapolate: Extrapolate.CLAMP,
  });
  const v = interpolate(animationValue, {
    inputRange,
    outputRange: colorsAsHSV.map(c => c.v),
    extrapolate: Extrapolate.CLAMP,
  });
  return hsv2color(h, s, v);
};

const interpolateColorsRGB = (
  animationValue: Animated.Adaptable<number>,
  inputRange: readonly Animated.Adaptable<number>[],
  colors: number[]
) => {
  const r = round(
    interpolate(animationValue, {
      inputRange,
      outputRange: colors.map(c => red(c)),
      extrapolate: Extrapolate.CLAMP,
    })
  );
  const g = round(
    interpolate(animationValue, {
      inputRange,
      outputRange: colors.map(c => green(c)),
      extrapolate: Extrapolate.CLAMP,
    })
  );
  const b = round(
    interpolate(animationValue, {
      inputRange,
      outputRange: colors.map(c => blue(c)),
      extrapolate: Extrapolate.CLAMP,
    })
  );
  const a = interpolate(animationValue, {
    inputRange,
    outputRange: colors.map(c => opacity(c)),
    extrapolate: Extrapolate.CLAMP,
  });

  return color(r, g, b, a);
};

export interface ReanimatedPath {
  totalLength: number;
  segments: { start: number; end: number; p0x: number; p3x: number }[];
  length: number[];
  start: number[];
  end: number[];
  p0x: number[];
  p0y: number[];
  p1x: number[];
  p1y: number[];
  p2x: number[];
  p2y: number[];
  p3x: number[];
  p3y: number[];
}

type BezierPoint =
  | 'p0x'
  | 'p0y'
  | 'p1x'
  | 'p1y'
  | 'p2x'
  | 'p2y'
  | 'p3x'
  | 'p3y';

export interface PathInterpolationConfig {
  inputRange: readonly Animated.Adaptable<number>[];
  outputRange: readonly (ReanimatedPath | string)[];
  extrapolate?: Animated.Extrapolate;
  extrapolateLeft?: Animated.Extrapolate;
  extrapolateRight?: Animated.Extrapolate;
}

export const interpolatePath = (
  value: Animated.Adaptable<number>,
  { inputRange, outputRange, ...config }: PathInterpolationConfig
): Animated.Node<string> => {
  const paths = outputRange.map(path =>
    typeof path === 'string' ? parsePath(path) : path
  );
  const [path] = paths;
  const commands = path.segments.map((_, index) => {
    const interpolatePoint = (point: BezierPoint) =>
      interpolate(value, {
        inputRange,
        outputRange: paths.map(p => p[point][index]),
        ...config,
      });

    const mx = interpolatePoint('p0x');
    const my = interpolatePoint('p0y');

    const p1x = interpolatePoint('p1x');
    const p1y = interpolatePoint('p1y');

    const p2x = interpolatePoint('p2x');
    const p2y = interpolatePoint('p2y');

    const p3x = interpolatePoint('p3x');
    const p3y = interpolatePoint('p3y');

    return string`${
      index === 0 ? string`M${mx},${my} ` : ''
    }C${p1x},${p1y} ${p2x},${p2y} ${p3x},${p3y}`;
  });
  return concat(...commands);
};

type SVGMoveCommand = ['M', number, number];
type SVGCurveCommand = ['C', number, number, number, number, number, number];

interface Point {
  x: number;
  y: number;
}

interface BezierCubicCurve {
  length: number;
  p0: Point;
  p1: Point;
  p2: Point;
  p3: Point;
}

export type Concatable =
  | Animated.Adaptable<string>
  | Animated.Adaptable<number>;

type SVGNormalizedCommands = [SVGMoveCommand, ...SVGCurveCommand[]];

export const string = (
  strings: readonly string[],
  ...values: readonly Concatable[]
) => {
  if (values.length === 0) {
    return concat(strings[0]);
  }
  const result = values.reduce<Concatable[]>(
    (acc, v, idx) => [...acc, strings[idx], v],
    []
  );
  result.push(strings[strings.length - 1]);
  return concat(...result);
};

const MX = 1;
const MY = 2;
const CX1 = 1;
const CY1 = 2;
const CX2 = 3;
const CY2 = 4;
const CX = 5;
const CY = 6;

export const parsePath = (d: string): ReanimatedPath => {
  const [move, ...curves]: SVGNormalizedCommands = normalizeSVG(
    absSVG(parseSVG(d))
  );
  // @ts-ignore
  const parts: BezierCubicCurve[] = curves.map((curve, index) => {
    const prevCurve = curves[index - 1];
    const p0 =
      index === 0
        ? { x: move[MX], y: move[MY] }
        : { x: prevCurve[CX], y: prevCurve[CY] };
    const p1 = { x: curve[CX1], y: curve[CY1] };
    const p2 = { x: curve[CX2], y: curve[CY2] };
    const p3 = { x: curve[CX], y: curve[CY] };
    const length = cubicBezierLength(p0, p1, p2, p3);
    return {
      p0,
      p1,
      p2,
      p3,
      length,
    };
  });
  const segments = parts.map((part, index) => {
    const start = parts.slice(0, index).reduce((acc, p) => acc + p.length, 0);
    const end = start + part.length;
    return {
      start,
      end,
      p0x: part.p0.x,
      p3x: part.p3.x,
    };
  });
  return {
    segments,
    totalLength: parts.reduce((acc, part) => acc + part.length, 0),
    length: parts.map(part => part.length),
    start: segments.map(segment => segment.start),
    end: segments.map(segment => segment.end),
    p0x: parts.map(part => part.p0.x),
    p0y: parts.map(part => part.p0.y),
    p1x: parts.map(part => part.p1.x),
    p1y: parts.map(part => part.p1.y),
    p2x: parts.map(part => part.p2.x),
    p2y: parts.map(part => part.p2.y),
    p3x: parts.map(part => part.p3.x),
    p3y: parts.map(part => part.p3.y),
  };
};
